File: mobile/library/kotlin/src/io/envoyproxy/envoymobile/StreamEmitter.kt
package io.envoyproxy.envoymobile

import java.nio.ByteBuffer

/**
 * Interface for a stream that may be canceled.
 */
interface CancelableStream {
  /**
   * Cancel and end the associated stream.
   */
  fun cancel()
}

/**
 * Interface allowing for sending/emitting data on an Envoy stream.
 */
interface StreamEmitter : CancelableStream {
  /**
   * For sending data to an associated stream.
   *
   * @param byteBuffer the byte buffer data to send to the stream.
   * @return this stream emitter.
   */
  fun sendData(byteBuffer: ByteBuffer): StreamEmitter

  /**
   * Close the stream with trailers.
   *
   * @param trailers trailers with which to close the stream.
   */
  fun close(trailers: Map<String, List<String>>)

  /**
   * Close the stream with a data frame.
   *
   * @param byteBuffer data with which to close the stream.
   */
  fun close(byteBuffer: ByteBuffer)
}

File: mobile/library/swift/src/StreamEmitter.swift
import Foundation

/// Interface for a stream that may be canceled.
@objc
public protocol CancelableStream {
  /// Cancel and end the associated stream.
  func cancel()
}

/// Interface allowing for sending/emitting data on an Envoy stream.
@objc
public protocol StreamEmitter: CancelableStream {
  /// Send data over the associated stream.
  ///
  /// - parameter data: Data to send over the stream.
  ///
  /// - returns: The stream emitter, for chaining syntax.
  @discardableResult
  func sendData(_ data: Data) -> StreamEmitter

  /// Close the stream with trailers.
  ///
  /// - parameter trailers: Trailers with which to close the stream.
  func close(trailers: [String: [String]])

  /// Close the stream with a data frame.
  ///
  /// - parameter data: Data with which to close the stream.
  func close(data: Data)
}
